syntax = "proto2";

package payload;

message packet {
  required fixed64 time_stamp = 1;

  enum MessageTypes {
      PEER_INFO = 0; 
      ADVERTISE = 1; 
      CHALLENGE = 2; 
      RESPONSE = 3; 
      SUCCESS = 4;
  }

  message Payload {
      required MessageTypes type = 1 [default = PEER_INFO];
      optional PeerInfo peerInfo = 2;
      optional Crypto crypto = 3; 
      optional OtherPeerInfo otherPeerInfo = 4;
  }

  message PeerInfo {
      required string ipAddress = 1; 
      required uint32 port = 2; 
      required string userName = 3; 
      required string peerName = 4; 
  }

  message Crypto {
      optional bytes hashedKey = 1; 
      optional bytes peerHashedKey = 2; 
      optional bytes encryptedMsg = 3;
      optional string nonce = 4;
      optional bytes encryptedNonce = 5;
      optional bytes encryptedFile = 6;   
  }

  message OtherPeerInfo {
      required uint32 peerPort = 1; 
      required string peerIpAddress = 2;
      required bool hasAdvertisedFirst = 3;  
  }

  required Payload payload = 2; 
}